### ISSUE TITLE
Get packet id for publish

### ISSUE THREAD

**Issue by shikoku1** [bytebeamio/rumqtt#349]
There doesn't seem to be a way to get the packet id for a published QoS1/2 message, to track its state
Is this correct? Is there a workaround?

**Comment by tekjar** [comment]
@shikoku1 In release R12 (tentative release in 2 weeks), `publish` method will return packet id on success

**Comment by mladedav** [comment]
@tekjar @de-sh does in-pipeline mean that there is a PR for this already? I couldn't find anything so I assume this was not released in March, correct?

I have looked around a bit and without changing the structure of client <-> envetloop architecture (where the latter is responsible for choosing packet IDs) I don't think there are other options than #291.

**Comment by tekjar** [comment]
`in-pipeline` means we've planned to work on this. This being worked in mqtt v5  PR where we use locks to access eventloop buffers. This allows us to return packet ids in client

**Comment by mladedav** [comment]
Is it planned also for 3.1.1? I've noticed that the v5 PR explicitly states that it doesn't make changes to lower versions.

I cannot migrate to 5 because of a broker but I would like to know that a message was delivered.

**Comment by tekjar** [comment]
Yeah also planned in 3.1.1. It's just that all the feedback we've collected is going into v5

**Comment by mladedav** [comment]
Ok, thanks for the clarifications.

**Comment by mladedav** [comment]
Hi @tekjar . I have noticed that R13 was released with the MQTT5 support, but the publish method still returns `Result<(), _>`.

Was this pushed further back? This is still planned for 3.1.1, right?

**Comment by tekjar** [comment]
Yeah this is planned @mladedav . Publish will return pkid. We still need to work on APIs

**Comment by svet-b** [comment]
Would very much love to see this implemented also. I noticed that in a few previous versions this was available for the v5 client, but was reverted more recently (due to other architectural changes).

Are there any workarounds in the meantime? Calling `publish()` and then inspecting the ID of the `Publish` event that's emitted on the loop? (seems brittle, of course, but maybe better than nothing)

**Comment by harlem88** [comment]
Is there any news about `Publish will return pkid` on `v3.1.1`?

**Comment by dlips** [comment]
Hi, I would also need this functionality. Is there any progress on this topic?

**Comment by dlips** [comment]
I have to send messages to a broker in a reliable way. To this end, I implemented a persistent message queue with rusqlite. Then I send the messages to a MQTT broker with QOS 1 and would like to remove the messages from the queue/db if I get the ACK for them.

**Comment by cartertinney** [comment]
I also need this. I need the ability to track when an ack for a specific publish is received.

**Comment by leoedin** [comment]
We can't see a way to implement functional QoS 1 or QoS 2 without being able to match sent packets with the received PubAck ID. This would give us one way of doing that.

If there's complications with locks, could you instead allow the user to pass in something when calling .publish() which is then passed to the event loop via the `Event::Outgoing::Publish` enum? The event loop handler would then have both the user token and pkid. I think that change can be implemented with fairly minimal code changes.

**Comment by Indy2222** [comment]
It should be possible to use an atomic for pkid counter stored in an Arc. The modular incrementation would be done with  [AtomicU16::fetch_update](https://doc.rust-lang.org/std/sync/atomic/struct.AtomicU16.html#method.fetch_update). The whole implementation would be lock-free and look like this:

```rust
let max_inflight = self.max_inflight;
self.pkid.fetch_update(Ordering::SeqCst, Ordering::Acquire, |prev| prev + 1 % max_inflight)
```

This would require pkid to be moved to AsyncClient and stored behind Arc. Also, the client would have to know `max_inflight` and have that in sync with the eventloop state (should be ok since the value never changes).

If you agree, I will submit a PR.

**Comment by tekjar** [comment]
We are working on an architecture where client calls like `client.publish()`, `client.subscribe()` returns a token which can be (a)waited on
